### ISSUE TITLE
Combine of covarince/contravariance delegates

### ISSUE THREAD

**Issue by gafter** [dotnet/coreclr#1712]
The delegate combine operator does not properly account for delegate variance.

See https://github.com/dotnet/roslyn/issues/3711 for details.

**Comment by gafter** [comment]
The managed language design team requests that this be addressed for the next VS product release.

**Comment by gafter** [comment]
@LLITCHEV: Can you please suggest who I should tag to get some commitment on this?

**Comment by LLITCHEV** [comment]
@jkotas Jan, could you please help Neal with the right people to talk to. I assume that this will require changes in the VM. Not so much in the Jit, but maybe yes?

**Comment by jkotas** [comment]
cc @gkhanna79 This is type system issue.

I do not think that it is a simple bug fix. We would need a new overload of Combine method that takes the type of the combined delegate as an argument.
 
Consider this case - there is no way for the existing Combine method to figure out that the result type should be `Func<IFace, object>`:

```
internal class Program
{
    interface IBaseA { }
    interface IBaseB { }

    interface IFace1 : IBaseA, IBaseB { }
    interface IFace2 : IBaseA, IBaseB { }

    interface IFace : IFace1, IFace2 { }

    static void Main(string[] args)
    {
        Func<IFace1, object> d1 = SomeFunction1;
        Func<IFace2, object> d2 = SomeFunction2;

        Func<IFace, object> e1 = d1;
        Func<IFace, object> e2 = d2;

        var d = e1 + e2;
    }

    static object SomeFunction1(IFace1 input)
    {
        return null;
    }

    static object SomeFunction2(IFace2 input)
    {
        return null;
    }
}
```

**Comment by gafter** [comment]
The C# compiler team will be happy to use whatever API you need to add.

**Comment by gkhanna79** [comment]
@JohnChen0 Can you please look into this for RC2?

**Comment by gafter** [comment]
@fadimounir @JohnChen0 @gkhanna79 Do you anticipate having this fixed in time to ship with C#7? The C# and VB compiler teams are happy to do our end of the work, but we need you to tell us which API we should be using.

**Comment by gkhanna79** [comment]
@gafter What is the timeframe for C# 7? We are heads down with .NET Core RTM and it will be good to understand the timeline before commenting on the next steps.

**Comment by gafter** [comment]
@gkhanna79 The time frame for getting this fixed in C# 7 is the same as the time frame for the ValueTuple libraries. Having this in place in the next month would enable us to fix the compiler for Dev15.

**Comment by gkhanna79** [comment]
@gafter I think we can look into this for .NET Core 1.2 as part of the overall API edition effort (it will be a Dev15 Update and not RTM).

**Comment by gafter** [comment]
@gkhanna79 We can have the compiler test for the presence of the new delegate combine API so that the fix will kick in for anyone who compiles against a platform that contains it. Please loop me in when it is being designed.

**Comment by gkhanna79** [comment]
@gafter @JohnChen0 will sync with you when he starts looking into this.

**Comment by gafter** [comment]
Ping. Just a reminder that the front ends [need this to produce correct code](https://github.com/dotnet/roslyn/issues/3711), and would be happy to work with you at your earliest convenience to get this addressed.

**Comment by gkhanna79** [comment]
@rahku Can you please look into this and determine the next steps?

**Comment by gafter** [comment]
Ping. Just a reminder that the front ends [need this to produce correct code](https://github.com/dotnet/roslyn/issues/3711), and would be happy to work with you at your earliest convenience to get this addressed.

**Comment by jkotas** [comment]
@MeiChin-Tsai will chat with @jaredpar about this and similar issues.

**Comment by gafter** [comment]
@jkotas Excellent. For reference a useful umbrella issue (a list of places where runtime bugs interfere with compiler correctness or language evolution) is https://github.com/dotnet/roslyn/issues/13295

**Comment by gkhanna79** [comment]
@jaredpar Is this being targeted for 2.0? If not, can you please move this out?

**Comment by TylerBrinkley** [comment]
Perhaps if [csharplang#103](https://github.com/dotnet/csharplang/issues/103) is implemented then this new generic covariant/contravariant `Delegate.Combine` API could be added.

**Comment by gafter** [comment]
I don't believe a new language feature is needed to address this. As @jkotas says

> We would need a new overload of `Combine` method that takes the type of the combined delegate as an argument

I think that is all that is needed.

**Comment by gafter** [comment]
Correction: we need a similar accommodation for `Delegate.Remove`. See https://github.com/dotnet/roslyn/issues/3711 for the original bug report, which demonstrates issues with both.

**Comment by RussKeldorph** [comment]
Future per @sergiy-k

**Comment by gafter** [comment]
We are still interested in getting these small APIs added so that the C# compiler can produce correct code.

**Comment by sergiy-k** [comment]
@MichalStrehovsky, could you please work with @gafter to define the exact set of APIs needed by the C# compiler and the cost of adding them? Thank you!

**Comment by gafter** [comment]
> The code currently works, and produces an instance of `Func<IFace1, object>` in `d` (while statically typed as `Func<IFace, object>` in C#).

This code does not work in some cases.

``` c#
using System;
public class C {
    public static void Main() {
        Func<IFace1, object> d1 = SomeFunction1;
        Func<IFace1, object> d2 = SomeFunction2;

        Func<IFace, object> e1 = d1;
        Func<IFace, object> e2 = d2;

        var d = e1 + e2;
    }
    
    static Func<IFace1, object> SomeFunction1 = SomeFunction1Impl;
    static object SomeFunction1Impl(IFace1 o) => o;
    static Func<IFace2, object> SomeFunction2 = SomeFunction2Impl;
    static object SomeFunction2Impl(IFace2 o) => o;
}

interface IFace2 { }
interface IFace1 : IFace2 { }
interface IFace : IFace1 { }
```

The new API should use the runtime type of the delegate if they are the same, reproducing the current behavior.  Only if they are not the same (in which case the current API would fail) would the new API produce a delegate of the provided runtime type.  This would be compatible.  In either case it would produce an object of the provided static type.
